src/dataflow/src/render/reduce.rs

```diff
@@ -114,34 +133,33 @@ where
                     |_expr| None,
                     move |row| {
                         let temp_storage = RowArena::new();
-                        // Ensure the packer is clear, and does not reflect
-                        // columns from prior rows that may have errored.
-                        row_packer.clear();
 
-                        // First, evaluate the key selector expressions.
+                        // Unpack only the demanded columns.
                         let mut datums_local = datums.borrow();
-                        datums_local.extend(row.iter().take(columns_needed));
-                        for expr in group_key_clone.iter() {
-                            match expr.eval(&datums_local, &temp_storage) {
-                                Ok(val) => row_packer.push(val),
-                                Err(e) => {
-                                    return Some(Err(e.into()));
-                                }
-                            }
+                        let mut row_iter = row.iter();
+                        for skip in demand.iter() {
+                            datums_local.push((&mut row_iter).nth(*skip).unwrap());
                         }
 
-                        // Second, evaluate the value selector expressions.
-                        let key = row_packer.finish_and_reuse();
-                        for aggr in aggregates_clone.iter() {
-                            match aggr.expr.eval(&datums_local, &temp_storage) {
-                                Ok(val) => {
-                                    row_packer.push(val);
-                                }
-                                Err(e) => {
-                                    return Some(Err(e.into()));
-                                }
-                            }
-                        }
+                        // Evaluate the key expressions.
+                        row_packer.clear();
+                        let key = match key_mfp.evaluate(
+                            &mut datums_local,
+                            &temp_storage,
+                            &mut row_packer,
+                        ) {
+                            Err(e) => return Some(Err(DataflowError::from(e))),
+                            Ok(Some(key)) => key,
+                            _ => panic!("Row expected as no predicate was used"),
```

Comment:
the panic kind of scares me just from a code hygiene pov could we maybe return a dataflow error instead? alternatively spitballing: maybe we could expose a `eval_map` function in mfp?

this is a relatively minor nit

Reply:
I think a dataflow error is probably wrong. This is a moment where within this code we can ensure that there is no predicate applied, so no opportunity to not produce a row. It doesn't bother me any more than any other `unwrap()` on a `Some(thing)`, except for the distance involved. Will ponder a bit.

Reply:
ah ok -- then I think I'm fine with

diff --git a/src/dataflow/src/render/reduce.rs b/src/dataflow/src/render/reduce.rs
index 846e01dd6..f80d82a0a 100644
--- a/src/dataflow/src/render/reduce.rs
+++ b/src/dataflow/src/render/reduce.rs

@@ -149,8 +149,7 @@ where
                             &mut row_packer,
                         ) {
                             Err(e) => return Some(Err(DataflowError::from(e))),
-                            Ok(Some(key)) => key,
-                            _ => panic!("Row expected as no predicate was used"),
+                            Ok(key) => key.expect("row known to exist as no predicate was used"),
                         };
                         // Evaluate the value expressions.
                         datums_local.truncate(demand.len());
```

thats a pretty minor change but i think its nice to not have raw `panic!` statements

Reply:
Can do. I think the logic used to be in a different order but your edit def improves the clarity.